Makefile

```diff
@@ -1,66 +1,151 @@
-		$(RES_DIR)/002bmc-apalache-0.6.0.csv \
-		$(RES_DIR)/002bmc-apalache-0.5.2.csv
+
+#################
+# PHONY TARGETS #
+#################
+
+.PHONY: reports experiments docker-pull prepare clean
+
+###########
+# REPORTS #
+###########
+
+# Reports aggregate data from all the benchmark data into human readible
+# summaries
+
+reports: $(foreach s, $(STRATEGIES), $(RES_DIR)/$(s)-report.md)
+
+$(RES_DIR)/%-report.md: $(call strategy_results,%)
 	cd ./results && \
-		$(BASEDIR)/scripts/mk-report.sh $(BASEDIR)/performance/002bmc-apalache.csv $^ >$@
+		$(BASEDIR)/scripts/mk-report.sh $(BASEDIR)/performance/$*-apalache.csv $^ >$@
+
+###############
+# EXPERIMENTS #
+###############
 
-# can we avoid duplication between 02bmc-apalache and 01indinv-apalache?
-$(RES_DIR)/001indinv-apalache-%.csv: prepare apalache-%
-	$(eval $@_NAME=001indinv-apalache-$*) # set the temporary variable
-	$(BASEDIR)/scripts/mk-run.py ./performance/001indinv-apalache.csv \
-		$(BUILDS_DIR)/apalache-$* ./performance $(RUN_DIR)/$($@_NAME)
-	(cd $(RUN_DIR)/$($@_NAME) && ./run-parallel.sh && \
-		$(BASEDIR)/scripts/parse-logs.py . && \
-		cp results.csv $(RES_DIR)/$($@_NAME).csv)
+# experiment-<strategy>-<version>:
+#
+# For each specified strategy and version, this loop generates a target
+# experiment-<strategy>-<version> to run the designated experiments.
+#
+# E.g. to run inductive invariant experiments for version 0.7.7 execute
+#
+# 	make experiment-001indinv-0.7.0
+$(foreach s, $(STRATEGIES), \
+$(foreach v, $(VERSIONS), \
+$(eval $(call experiment-strat-version,$(s),$(v)))))
 
-# can we avoid duplication between 02bmc-apalache and 01indinv-apalache?
-$(RES_DIR)/002bmc-apalache-%.csv: prepare apalache-%
-	$(eval $@_NAME=002bmc-apalache-$*) # set the temporary variable
-	$(BASEDIR)/scripts/mk-run.py ./performance/002bmc-apalache.csv \
-		$(BUILDS_DIR)/apalache-$* ./performance $(RUN_DIR)/$($@_NAME)
-	(cd $(RUN_DIR)/$($@_NAME) && ./run-parallel.sh && \
-		$(BASEDIR)/scripts/parse-logs.py . && \
-		cp results.csv $(RES_DIR)/$($@_NAME).csv)
+# Run all experiments for all versions
+experiments: $(foreach s, $(STRATEGIES), $(foreach v, $(VERSIONS), experiment-$(s)-$(v)))
 
-build: apalache-0.5.2 apalache-0.6.0 apalache-0.7.0
+# Rules for generating the csv of result data for a particular set of
+# experimen
+#
+# The pattern % will look like <strategy>-apalache-<version>
+# for the given STRATEGY and VERSION
+$(RES_DIR)/%.csv: prepare docker-pull
+	@echo
+	@echo "======> Running experiments " $*
+	@echo
+# PARAMS is the base name for the params file, obtained by stripping the
+# version segment from the end of the filename.
+# NOTE: The double $ in the sed command escapes the $ for make
+	$(eval PARAMS := $(shell echo $* | sed "s/-[^-]*$$//"))
+	$(eval VERSION := $(shell echo $* | sed 's/.*-\(.*\)$$/\1/'))
+	$(BASEDIR)/scripts/mk-run.py \
+		./performance/$(PARAMS).csv \
+		$(VERSION) \
+		./performance \
+		$(RUN_DIR)/$*
+	(cd $(RUN_DIR)/$* \
+		&& ./run-parallel.sh\
```

Comment:
could you introduce an extra parameter that allows the user to choose between `run-parallel` and `run-all`, as all experiments run in parallel may cause memory congestion in the future.

Reply:
Good idea!

For the future, I think the "right" way to go here would be to use a build system that is more expressive and will take care of the parallelism etc for us. From what I understand of the whole set up here, it call all be viewed as a build problem: given some source files and some dependencies, we want to generate some artifacts: csv and markdown files. Last week I experimented a little bit with using Bazel for this purpose, and I'm pretty sure it would work well and give us some nice properties. The cost, however, would be a pretty deep overhaul. -- Just something to consider for the future :)

Reply:
Addressed in https://github.com/informalsystems/apalache-tests/pull/9/commits/240228b4bb8eb57d9f2173a409388457fd39ee73 let me know if this gives us what we want here.

Reply:
Yes, that looks nice. I am not sure about treating the experiments as a build problem. When you build something, the resource consumption is more or less predictable and preemption is no problem. For the experiments, we do not want different benchmarks to compete with each other for resources. So the serial execution is the right way. In the past, I was running experiments in a cluster, so it was not an issue.

Reply:
Ah yes, I follow you there.

So, with a bit more refactoring, I think it should be possible to actually have a distinct target for every experiment (rather than just for the experiment suites, like we currently have), this would let us configure the CI build matrix to run every experiment in it's own machine. IIUC, that would be desirable, and let us have parallelism (up the capacity of our infrastructure to run machines in tandem) without risking memory congestion. Does it make sense for us to keep this in mind for a future optimization?

For the time being, however, I chink I can just configure the CI to run each suite sequentially, following your guidelines here.